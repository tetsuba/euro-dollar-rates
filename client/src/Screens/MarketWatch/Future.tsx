import React from 'react'
import service from '../../utils/service'

// HOC
import { withRouter } from '../../HOC/withRouter'

// COMPONENTS
import BreadCrumbs from '../../Components/BreadCrumbs/BreadCrumbs'
import LineChart from '../../Components/Charts/LineChart'
import FutureCheckBoxList from '../../Components/List/FutureCheckBoxList'

// UTILS
import {
    mutateFutureList,
    mutateFutureListForLineChart,
    mutateFutureStringToObject,
} from '../../utils/mutations'
import { EventClickType, FutureListType, FutureType } from '../../utils/types'
import LoadingSpinner from '../../Components/LoadingSpinner/LoadingSpinner'

interface PropTypes {
    params: {
        [key: string]: string
    }
}


interface StateTypes {
    future: {
        data: {
            list: FutureType[]
        }
    }
    futureList: FutureType[]
    futureYear: FutureListType[]
    showHideLineChart: boolean
}


class Future extends React.Component<PropTypes, StateTypes> {
    state = {
        future: {
            data: {
                list: [],
            },
        },
        futureList: [],
        futureYear: [],
        showHideLineChart: false,
    }

    async componentDidMount() {
        const future = await service.getFuture(this.props.params.symbol)

        this.setState({
            future: future.data,
            futureList: mutateFutureList(future.data.list),
            futureYear: future.data.list.map(mutateFutureStringToObject),
        })
    }

    handlerOnClickCheckbox = (e: EventClickType): void => {
        const { futureList } = this.state
        const el = e.target as HTMLElement
        const index: null | string = el.getAttribute('data-index')
        const checked: null | string = el.getAttribute('data-checked')

        // TODO: Look into this and find out why TS is producing this error
        // TS2339: Property 'checked' does not exist on type 'never'.
        // @ts-ignore
        futureList[Number(index)].checked = !(checked === 'true')
        // futureList[Number(index)].checked = !(checked === 'true')
        this.setState({ futureList })
    }

    handlerOnClickButton = () => {
        this.setState({
            showHideLineChart: !this.state.showHideLineChart,
        })
    }

    renderChart(futureList: FutureType[]) {
        return (
            <>
                <div className="row">
                    <button
                        type="button"
                        className="btn btn-primary"
                        onClick={this.handlerOnClickButton}
                    >
                        Back
                    </button>
                </div>
                <div className="row mt-5">
                    <LineChart {...mutateFutureListForLineChart(futureList)} />
                </div>
            </>
        )
    }

    renderList(futureList: FutureType[]) {
        return (
            <>
                <div className="row">
                    <button
                        type="button"
                        className="btn btn-primary"
                        onClick={this.handlerOnClickButton}
                    >
                        Create Chart
                    </button>
                </div>
                <div className="row mt-5" data-testid="future-checkbox-list">
                    <FutureCheckBoxList
                        list={futureList}
                        onClick={this.handlerOnClickCheckbox}
                    />
                </div>
            </>
        )
    }

    render() {
        const { showHideLineChart, futureList } = this.state

        return (
            <>
                <BreadCrumbs />
                {futureList.length < 1 ? (
                    <LoadingSpinner />
                ) : showHideLineChart ? (
                    this.renderChart(futureList)
                ) : (
                    this.renderList(futureList)
                )}
            </>
        )
    }
}

export default withRouter(Future)
