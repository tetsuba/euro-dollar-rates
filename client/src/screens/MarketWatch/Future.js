import React from 'react'
import service from '../../utils/service'
import { withRouter } from '../../HOC/withRouter'
import BreadCrumbs from '../../components/BreadCrumbs/BreadCrumbs'
import LineChart from '../../components/Charts/LineChart'
import {
    mutateFutureList,
    mutateFutureListForLineChart,
    mutateFutureStringToObject,
} from '../../utils/mutations'
import FutureCheckBoxList from '../../components/List/FutureCheckBoxList';

class Future extends React.Component {
    state = {
        future: {},
        showHideLineChart: false,
    }

    async componentDidMount() {
        const future = await service.getFuture(this.props.params.symbol)

        this.setState({
            future: future.data,
            futureList: mutateFutureList(future.data.list),
            futureYear: future.data.list.map(mutateFutureStringToObject),
        })
    }

    handlerOnClickCheckbox = (e) => {
        const { futureList } = this.state
        const index = e.target.getAttribute('index')
        const checked = e.target.getAttribute('data-checked')
        futureList[index].checked = !(checked === 'true')
        this.setState({
            futureList,
        })
    }

    handlerOnClickButton = () => {
        this.setState({
            showHideLineChart: !this.state.showHideLineChart,
        })
    }

    render() {
        const { showHideLineChart, futureList } = this.state

        return (
            <>
                <BreadCrumbs />
                {showHideLineChart && (
                    <>
                        <div className="row">
                            <button
                                type="button"
                                className="btn btn-primary"
                                onClick={this.handlerOnClickButton}
                            >
                                Back
                            </button>
                        </div>
                        <div className="row mt-5">
                            <LineChart
                                data={mutateFutureListForLineChart(futureList)}
                            />
                        </div>
                    </>
                )}
                {futureList && !showHideLineChart && (
                    <>
                        <div className="row">
                            <button
                                type="button"
                                className="btn btn-primary"
                                onClick={this.handlerOnClickButton}
                            >
                                Create Chart
                            </button>
                        </div>
                        <div className="row mt-5" data-testid="future-checkbox-list">
                            <FutureCheckBoxList
                              list={futureList}
                              onClick={this.handlerOnClickCheckbox}
                            />
                        </div>
                    </>
                )}
            </>
        )
    }
}

export default withRouter(Future)
