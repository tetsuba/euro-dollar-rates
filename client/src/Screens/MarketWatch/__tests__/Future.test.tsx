import React from 'react'
import axios from 'axios'
import { fireEvent, render, screen, waitFor } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import Future from '../Future'
import 'jest-canvas-mock'

jest.mock('axios')

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useParams: () => ({
        symbol: 'ED',
    }),
}))

jest.mock('chart.js', () => {
    return jest.fn()
})

const mockData = {
    data: {
        list: [
            {
                year: 2021,
                month: '06',
                price: '1,000,000.50',
                name: 'Eurodollar 3 Month Jun 2021',
                symbol: 'EDM21',
            },
            {
                year: 2010,
                month: '06',
                price: '100,000',
                name: 'Eurodollar 3 Month Jun 2010',
                symbol: 'EDM10',
            },
            {
                year: 1999,
                month: '06',
                price: '1,000',
                name: 'Eurodollar 3 Month Jun 1999',
                symbol: 'EDM99',
            },
            {
                year: 2001,
                month: '06',
                price: '10,000',
                name: 'Eurodollar 3 Month Jun 2000',
                symbol: 'EDM01',
            },
        ],
    },
}

describe('Future', () => {
    // @ts-ignore
    axios.get.mockReturnValueOnce(mockData)

    const rendered = render(
        <MemoryRouter
            initialEntries={['', { pathname: '/market-watch/futures/ED' }]}
        >
            <Future />
        </MemoryRouter>
    )

    test('should render', async () => {
        await waitFor(() => expect(screen.getByTestId('future-checkbox-list')))
        expect(rendered.asFragment()).toMatchSnapshot()

        /* Event: onClick
         * Button: Create Chart
         * Test: should hide future checkbox list and display line chart
         *  */
        fireEvent.click(rendered.getByText(/Create Chart/i))
        await waitFor(() => expect(rendered.getByTestId('line-chart')))
        expect(rendered.queryByTestId('future-checkbox-list')).toBeNull()

        /* Event: onClick
         * Button: Back
         * Test: should hide line chart and display future checkbox list
         *  */
        fireEvent.click(rendered.getByText(/Back/i))
        await waitFor(() =>
            expect(rendered.getByTestId('future-checkbox-list'))
        )
        expect(rendered.queryByTestId('line-chart')).toBeNull()

        /* Test: all checkboxes should be checked = 'true'
         *  */
        rendered.queryAllByTestId('checkbox-custom').forEach((node) => {
            expect(node.getAttribute('data-checked')).toMatch('true')
        })

        /* Event: onClick
         * Checkbox: uncheck
         * Test: should uncheck the first checkbox in the list
         *  */
        const firstCheckBox = rendered.queryAllByTestId('checkbox-custom')[0]
        fireEvent.click(firstCheckBox)
        expect(firstCheckBox.getAttribute('data-checked')).toMatch('false')
    })
})
