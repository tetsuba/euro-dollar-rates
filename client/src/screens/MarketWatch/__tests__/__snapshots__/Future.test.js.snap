// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Future should render 1`] = `
<DocumentFragment>
  <div
    class="row shadow mb-5"
  >
    <nav
      aria-label="breadcrumb"
    >
      <ol
        class="breadcrumb mt-3"
      >
        <li
          class="breadcrumb-item"
        >
          <a
            href="/"
          >
            Home
          </a>
        </li>
        <li
          class="breadcrumb-item"
        >
          <a
            href="/market-watch"
          >
            market-watch
          </a>
        </li>
        <li
          class="breadcrumb-item"
        >
          <a
            href="/market-watch/futures"
          >
            futures
          </a>
        </li>
        <li
          aria-current="page"
          class="breadcrumb-item active"
        >
          ED
        </li>
      </ol>
    </nav>
  </div>
  <div
    class="row"
  >
    <button
      class="btn btn-primary"
      type="button"
    >
      Create Chart
    </button>
  </div>
  <div
    class="row mt-5"
    data-testid="future-checkbox-list"
  >
    <ul
      class="list-group list-group-flush"
    >
      <li
        class="list-group-item"
      >
        <div
          class="form-check"
          data-checked="true"
          data-testid="checkbox-custom"
          index="0"
        >
          <input
            checked=""
            class="form-check-input pe-none"
            id="flexCheckDefault"
            type="checkbox"
          />
          <label
            class="form-check-label pe-none"
            for="flexCheckDefault"
          >
            1999/06 - 1,000
          </label>
        </div>
      </li>
      <li
        class="list-group-item"
      >
        <div
          class="form-check"
          data-checked="true"
          data-testid="checkbox-custom"
          index="1"
        >
          <input
            checked=""
            class="form-check-input pe-none"
            id="flexCheckDefault"
            type="checkbox"
          />
          <label
            class="form-check-label pe-none"
            for="flexCheckDefault"
          >
            2001/06 - 10,000
          </label>
        </div>
      </li>
      <li
        class="list-group-item"
      >
        <div
          class="form-check"
          data-checked="true"
          data-testid="checkbox-custom"
          index="2"
        >
          <input
            checked=""
            class="form-check-input pe-none"
            id="flexCheckDefault"
            type="checkbox"
          />
          <label
            class="form-check-label pe-none"
            for="flexCheckDefault"
          >
            2010/06 - 100,000
          </label>
        </div>
      </li>
      <li
        class="list-group-item"
      >
        <div
          class="form-check"
          data-checked="true"
          data-testid="checkbox-custom"
          index="3"
        >
          <input
            checked=""
            class="form-check-input pe-none"
            id="flexCheckDefault"
            type="checkbox"
          />
          <label
            class="form-check-label pe-none"
            for="flexCheckDefault"
          >
            2021/06 - 1,000,000.50
          </label>
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
